File: ios/RunWrap/Theme.swift
import SwiftUI
import UIKit

/// 디자인 토큰 — 디자인 시안(claude.design "Runner Report" v0.4)의 rr-theme 팔레트.
/// 라이트/다크 값을 UIColor 다이내믹 프로바이더로 묶어 시스템 모드를 따른다.
/// v0.4는 "스포티 볼드" 리디자인 — 웜 크림 배경 + 레드오렌지 브랜드가 기준이다.
enum RR {
    static let bg = adaptive(0xF2F0EA, 0x0A0A09)
    static let surface = adaptive(0xFFFFFF, 0x151513)
    static let surface2 = adaptive(0xEAE7E0, 0x1E1E1B)
    static let text = adaptive(0x12120F, 0xF5F4EF)
    static let text2 = adaptive(0x54544E, 0xA8A79E)
    static let text3 = adaptive(0x8F8F86, 0x73736C)
    static let brand = adaptive(0xFF4D2E, 0xFF5A3C)
    static let pos = adaptive(0x0E9146, 0x35E077)
    /// 강수(비·눈) 심볼 전용 청색 — 웜 팔레트의 유일한 한랭 색. 날씨 아이콘 palette 렌더링에만 쓴다
    static let sky = adaptive(0x2E8BD9, 0x5AAEFF)
    static let warn = adaptive(0xC77700, 0xFFAE00)
    static let dang = adaptive(0xD91F00, 0xFF3B30)

    /// 기동 스플래시 전용 — 앱 아이콘의 파랑 그라데이션 중간 톤 (light #16A5FB / dark #101426).
    /// 런치 스크린(Info.plist UIColorName)이 같은 에셋을 참조하므로 hex가 아니라 에셋에서 온다 —
    /// 색을 바꾸려면 LaunchBackground.colorset 한 곳만 고친다
    static let launchBg = Color("LaunchBackground")

    /// 시안: light rgba(20,20,16,.13) / dark rgba(255,255,255,.11)
    static let line = alpha(black: 0.13, white: 0.11)
    /// 차트 막대 바탕 — light rgba(20,20,16,.09) / dark rgba(255,255,255,.10)
    static let barFill = alpha(black: 0.09, white: 0.10)

    static let brandSoft = soft(brand, 0.12, 0.18)
    static let posSoft = soft(pos, 0.12, 0.16)
    static let warnSoft = soft(warn, 0.13, 0.16)
    static let dangSoft = soft(dang, 0.11, 0.16)

    private static func adaptive(_ light: UInt32, _ dark: UInt32) -> Color {
        Color(UIColor { $0.userInterfaceStyle == .dark ? UIColor(rgb: dark) : UIColor(rgb: light) })
    }

    private static func alpha(black: CGFloat, white: CGFloat) -> Color {
        Color(UIColor {
            $0.userInterfaceStyle == .dark
                ? UIColor.white.withAlphaComponent(white)
                : UIColor.black.withAlphaComponent(black)
        })
    }

    private static func soft(_ base: Color, _ light: CGFloat, _ dark: CGFloat) -> Color {
        let ui = UIColor(base)
        return Color(UIColor {
            ui.resolvedColor(with: $0).withAlphaComponent($0.userInterfaceStyle == .dark ? dark : light)
        })
    }

    /// 화면 대제목용 디스플레이 서체 — 시안 v0.4의 'Black Han Sans'(번들 .ttf, SIL OFL 1.1).
    /// 한글 전용 굵은 서체라 획이 두꺼워 시안처럼 weight를 따로 주지 않는다.
    static func display(_ size: CGFloat) -> Font {
        .custom("BlackHanSans-Regular", size: size)
    }

    /// 큰 숫자용 디스플레이 서체 — 시안 v0.4의 'Anton'(번들 .ttf, SIL OFL 1.1).
    /// 라틴 숫자 전용이라 한글이 섞이는 자리에는 쓰지 않는다 (한글 글리프가 없어 폴백된다).
    static func numeral(_ size: CGFloat) -> Font {
        .custom("Anton-Regular", size: size)
    }
}

private extension UIColor {
    convenience init(rgb: UInt32) {
        self.init(red: CGFloat((rgb >> 16) & 0xFF) / 255,
                  green: CGFloat((rgb >> 8) & 0xFF) / 255,
                  blue: CGFloat(rgb & 0xFF) / 255,
                  alpha: 1)
    }
}

/// 카드 상태 톤 4가지 — 시안의 배지/강조색 매핑 (과부하·주의·유지·개선)
enum RRTone {
    case overload, caution, steady, improving

    var color: Color {
        switch self {
        case .overload: RR.dang
        case .caution: RR.warn
        case .steady: RR.brand
        case .improving: RR.pos
        }
    }

    var softColor: Color {
        switch self {
        case .overload: RR.dangSoft
        case .caution: RR.warnSoft
        case .steady: RR.brandSoft
        case .improving: RR.posSoft
        }
    }

    var label: String {
        switch self {
        case .overload: "과부하 구간"
        case .caution: "주의 구간"
        case .steady: "유지 중"
        case .improving: "좋아지는 중"
        }
    }

    var code: String {
        switch self {
        case .overload: "OVERLOAD"
        case .caution: "CAUTION"
        case .steady: "STEADY"
        case .improving: "IMPROVING"
        }
    }
}

/// 시안 v0.4의 상태 배지: 굵은 16×4 막대 + 한글 라벨 + 영문 코드.
/// v0.3까지의 soft 배경 알약에서 배경 없는 플랫 형태로 바뀌었다 — 카드 상단에서
/// 색 면적을 줄이고 헤드라인이 주인공이 되게 하려는 의도.
struct ToneBadge: View {
    let tone: RRTone
    /// 톤 라벨을 문맥에 맞게 덮어쓴다 (예: 칼로리 카드의 "BURNING", 꾸준함 카드의 "STREAK")
    var label: String?
    var code: String?

    var body: some View {
        HStack(spacing: 8) {
            Rectangle().fill(tone.color).frame(width: 16, height: 4)
            Text(label ?? tone.label)
                .font(.system(size: 11, weight: .heavy))
                .kerning(0.55)
                .foregroundStyle(tone.color)
            Text(code ?? tone.code)
                .font(.system(size: 10, weight: .semibold, design: .monospaced))
                .kerning(1)
                .foregroundStyle(tone.color.opacity(0.7))
        }
    }
}

/// 시안 v0.4의 기본 카드: surface + 1px line + r12 + 옅은 그림자.
/// v0.3의 큰 곡률(r24)에서 스포티한 각진 인상으로 조정됐다.
struct RRCardModifier: ViewModifier {
    var radius: CGFloat = 12

    func body(content: Content) -> some View {
        content
            .background(RR.surface, in: RoundedRectangle(cornerRadius: radius, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: radius, style: .continuous).strokeBorder(RR.line))
            .shadow(color: .black.opacity(0.04), radius: 1, y: 1)
    }
}

extension View {
    func rrCard(radius: CGFloat = 12) -> some View { modifier(RRCardModifier(radius: radius)) }
}

/// 실내(트레드밀) 세션 표시용 소형 텍스트 배지 — 상태(톤)가 아니라 종류 표시라
/// RRTone 매핑을 쓰지 않는다 (계획서 M1). 시안: 10px/650, brandSoft 배경
struct IndoorBadge: View {
    var body: some View {
        Text("실내")
            .font(.system(size: 10, weight: .semibold))
            .foregroundStyle(RR.brand)
            .padding(.horizontal, 6)
            .padding(.vertical, 2.5)
            .background(RR.brandSoft, in: RoundedRectangle(cornerRadius: 4, style: .continuous))
    }
}

/// 화면 상단 모노스페이스 아이브로 라벨 (예: "2026년 8월 2째주")
struct Eyebrow: View {
    let text: String

    var body: some View {
        Text(text.uppercased())
            .font(.system(size: 11, weight: .semibold, design: .monospaced))
            .kerning(1.4)
            .foregroundStyle(RR.text3)
    }
}

/// 수치 포맷 공통 헬퍼
enum Format {
    /// "1:52:34" 또는 "48:10"
    static func duration(_ seconds: Double) -> String {
        let s = Int(seconds.rounded())
        return s >= 3600
            ? String(format: "%d:%02d:%02d", s / 3600, (s % 3600) / 60, s % 60)
            : String(format: "%d:%02d", s / 60, s % 60)
    }

    /// "5′20″"
    static func pace(_ secPerKm: Double) -> String {
        let s = Int(secPerKm.rounded())
        return "\(s / 60)′\(String(format: "%02d", s % 60))″"
    }

    /// "5′20″/km"
    static func paceKm(_ secPerKm: Double) -> String { pace(secPerKm) + "/km" }

    static func km(_ value: Double) -> String { String(format: "%.1f", value) }

    /// 걷뛰 분 — 정수는 "2", 반 분은 "3.5". 사다리에 0.5분 단위가 있어 나눠 찍는다.
    /// 엔진의 `headline`과 카드의 metric이 같은 표기를 쓰도록 여기 둔다 (WalkRunEngine)
    static func walkRunMinutes(_ value: Double) -> String {
        value == value.rounded() ? String(format: "%.0f", value)
                                 : String(format: "%.1f", value)
    }

    /// "4,120" — 천 단위 구분 정수 (칼로리 카드)
    static func kcal(_ value: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? "0"
    }

    /// "8월 2째주" — 차트 주 라벨. 두 달에 걸친 주는 그 주의 목요일이 속한 달로
    /// 정하고(ISO 8601 관행), 째주 순번도 목요일 날짜 기준(1–7일 → 1째주)이다.
    /// withYear를 켜면 "2026년 8월 2째주" — 연도도 목요일 기준(연말·연초에 걸친 주 대응).
    static func weekLabel(weekStart: Date, withYear: Bool = false) -> String {
        var calendar = Calendar(identifier: .iso8601)  // 월요일 시작
        calendar.timeZone = .current
        let thursday = calendar.date(byAdding: .day, value: 3, to: weekStart) ?? weekStart
        let month = calendar.component(.month, from: thursday)
        let ordinal = (calendar.component(.day, from: thursday) + 6) / 7
        let label = "\(month)월 \(ordinal)째주"
        return withYear ? "\(calendar.component(.year, from: thursday))년 \(label)" : label
    }
}

File: android/app/src/main/java/com/jkpark/runwrap/ui/theme/Theme.kt
package com.jkpark.runwrap.ui.theme

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.jkpark.runwrap.R

/// 디자인 토큰 — iOS `Theme.swift`의 rr-theme 팔레트(claude.design "Runner Report" v0.4)를
/// 수치 그대로 이식. 라이트/다크 팔레트를 CompositionLocal로 내려 화면은 `RR.bg`처럼
/// iOS와 같은 표기로만 색을 읽는다 (화면 코드에서 색 리터럴 금지 — iOS 규칙 유지).
@Immutable
data class RRPalette(
    val bg: Color,
    val surface: Color,
    val surface2: Color,
    val text: Color,
    val text2: Color,
    val text3: Color,
    val brand: Color,
    val pos: Color,
    /// 강수(비·눈) 심볼 전용 청색 — 웜 팔레트의 유일한 한랭 색
    val sky: Color,
    val warn: Color,
    val dang: Color,
    /// 시안: light rgba(20,20,16,.13) / dark rgba(255,255,255,.11)
    val line: Color,
    /// 차트 막대 바탕 — light rgba(20,20,16,.09) / dark rgba(255,255,255,.10)
    val barFill: Color,
    val brandSoft: Color,
    val posSoft: Color,
    val warnSoft: Color,
    val dangSoft: Color,
)

private fun rgb(hex: Long) = Color(0xFF00_0000L or hex)

val LightRRPalette = RRPalette(
    bg = rgb(0xF2F0EA), surface = rgb(0xFFFFFF), surface2 = rgb(0xEAE7E0),
    text = rgb(0x12120F), text2 = rgb(0x54544E), text3 = rgb(0x8F8F86),
    brand = rgb(0xFF4D2E), pos = rgb(0x0E9146), sky = rgb(0x2E8BD9),
    warn = rgb(0xC77700), dang = rgb(0xD91F00),
    line = Color.Black.copy(alpha = 0.13f), barFill = Color.Black.copy(alpha = 0.09f),
    brandSoft = rgb(0xFF4D2E).copy(alpha = 0.12f), posSoft = rgb(0x0E9146).copy(alpha = 0.12f),
    warnSoft = rgb(0xC77700).copy(alpha = 0.13f), dangSoft = rgb(0xD91F00).copy(alpha = 0.11f),
)

val DarkRRPalette = RRPalette(
    bg = rgb(0x0A0A09), surface = rgb(0x151513), surface2 = rgb(0x1E1E1B),
    text = rgb(0xF5F4EF), text2 = rgb(0xA8A79E), text3 = rgb(0x73736C),
    brand = rgb(0xFF5A3C), pos = rgb(0x35E077), sky = rgb(0x5AAEFF),
    warn = rgb(0xFFAE00), dang = rgb(0xFF3B30),
    line = Color.White.copy(alpha = 0.11f), barFill = Color.White.copy(alpha = 0.10f),
    brandSoft = rgb(0xFF5A3C).copy(alpha = 0.18f), posSoft = rgb(0x35E077).copy(alpha = 0.16f),
    warnSoft = rgb(0xFFAE00).copy(alpha = 0.16f), dangSoft = rgb(0xFF3B30).copy(alpha = 0.16f),
)

val LocalRRPalette = staticCompositionLocalOf { LightRRPalette }

/// 화면에서 `RR.bg`처럼 iOS와 같은 표기로 읽는 접근자
object RR {
    val bg: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.bg
    val surface: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.surface
    val surface2: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.surface2
    val text: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.text
    val text2: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.text2
    val text3: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.text3
    val brand: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.brand
    val pos: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.pos
    val sky: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.sky
    val warn: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.warn
    val dang: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.dang
    val line: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.line
    val barFill: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.barFill
    val brandSoft: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.brandSoft
    val posSoft: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.posSoft
    val warnSoft: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.warnSoft
    val dangSoft: Color @Composable @ReadOnlyComposable get() = LocalRRPalette.current.dangSoft

    /// 디스플레이 서체 스타일 — iOS `RR.display(size:)` 대응 (한글 제목 전용)
    fun display(size: TextUnit) = TextStyle(fontFamily = RRDisplayFamily, fontSize = size)

    /// 큰 숫자 서체 스타일 — iOS `RR.numeral(size:)` 대응.
    /// Anton은 라틴 숫자 전용이라 한글 자리에 쓰지 않는다 (iOS와 같은 규칙).
    fun numeral(size: TextUnit) = TextStyle(fontFamily = RRNumeralFamily, fontSize = size)
}

/// 번들 서체 — iOS와 같은 .ttf(SIL OFL 1.1)를 res/font로 복사해 두었다
val RRDisplayFamily = FontFamily(Font(R.font.black_han_sans))
val RRNumeralFamily = FontFamily(Font(R.font.anton))

/// 앱 루트 테마 — 시스템 모드에 따라 RR 팔레트를 내려보낸다.
@Composable
fun RunWrapTheme(content: @Composable () -> Unit) {
    val palette = if (isSystemInDarkTheme()) DarkRRPalette else LightRRPalette
    CompositionLocalProvider(LocalRRPalette provides palette) {
        MaterialTheme(content = content)
    }
}

/// 카드 상태 톤 4가지 — 과부하·주의·유지·개선 (iOS `RRTone` 대응).
/// 엔진은 색이 아니라 이 톤까지만 결정한다. label/code는 순수 값이라 JVM 테스트 가능,
/// 색 매핑은 팔레트가 필요해 @Composable 접근자로만 연다.
enum class RRTone {
    OVERLOAD, CAUTION, STEADY, IMPROVING;

    val label: String
        get() = when (this) {
            OVERLOAD -> "과부하 구간"
            CAUTION -> "주의 구간"
            STEADY -> "유지 중"
            IMPROVING -> "좋아지는 중"
        }

    val code: String
        get() = when (this) {
            OVERLOAD -> "OVERLOAD"
            CAUTION -> "CAUTION"
            STEADY -> "STEADY"
            IMPROVING -> "IMPROVING"
        }

    val color: Color
        @Composable @ReadOnlyComposable get() = when (this) {
            OVERLOAD -> RR.dang
            CAUTION -> RR.warn
            STEADY -> RR.brand
            IMPROVING -> RR.pos
        }

    val softColor: Color
        @Composable @ReadOnlyComposable get() = when (this) {
            OVERLOAD -> RR.dangSoft
            CAUTION -> RR.warnSoft
            STEADY -> RR.brandSoft
            IMPROVING -> RR.posSoft
        }
}

/// 기본 카드 — iOS `.rrCard(radius:)` 대응: surface 배경 + 1dp line 테두리 +
/// 검정 4% 그림자, 모서리 r12.
fun Modifier.rrCard(radius: Dp = 12.dp): Modifier = composed {
    val shape = RoundedCornerShape(radius)
    shadow(
        elevation = 1.dp, shape = shape,
        ambientColor = Color.Black.copy(alpha = 0.04f),
        spotColor = Color.Black.copy(alpha = 0.04f),
    )
        .background(RR.surface, shape)
        .border(1.dp, RR.line, shape)
}

/// 톤 배지 — 색 대시 + 라벨 + 코드 (iOS `ToneBadge` 대응).
/// label/code를 넘기면 톤 기본 문안 대신 그 문안을 쓴다 (예: "BURNING"/"STREAK").
@Composable
fun ToneBadge(tone: RRTone, label: String? = null, code: String? = null) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.size(width = 16.dp, height = 4.dp).background(tone.color))
        Text(
            label ?: tone.label,
            fontSize = 11.sp, fontWeight = FontWeight.ExtraBold,
            letterSpacing = 0.55.sp, color = tone.color,
        )
        Text(
            code ?: tone.code,
            fontSize = 10.sp, fontWeight = FontWeight.SemiBold,
            fontFamily = FontFamily.Monospace, letterSpacing = 1.sp,
            color = tone.color.copy(alpha = 0.7f),
        )
    }
}

/// 섹션 아이브로 라벨 — 소문자를 넣어도 대문자로 보인다 (iOS `Eyebrow` 대응)
@Composable
fun Eyebrow(text: String) {
    Text(
        text.uppercase(),
        fontSize = 11.sp, fontWeight = FontWeight.SemiBold,
        fontFamily = FontFamily.Monospace, letterSpacing = 1.4.sp,
        color = RR.text3,
    )
}

/// 실내(트레드밀) 배지 — 목록·상세 공용 (iOS `IndoorBadge` 대응)
@Composable
fun IndoorBadge() {
    Text(
        "실내",
        Modifier
            .background(RR.brandSoft, RoundedCornerShape(4.dp))
            .padding(horizontal = 6.dp, vertical = 2.5.dp),
        fontSize = 10.sp, fontWeight = FontWeight.SemiBold,
        color = RR.brand,
    )
}
